### PR TITLE
Evals groundwork

### PR DESCRIPTION
**Pull request by valedan**
This is some supporting work and refactoring that I've broken out of my big upcoming evals PR to make things a little more manageable.

Included here:
- Unifying the interface for eval functions (pathdist.py -> path_evals.py). Minimal functional changes here, just renaming some stuff and letting `node_overlap` take numpy arrays. These are not unit tested but #112 is tracking that.
- Move constants to seperate file
- Add token utils and move `decode_maze_tokens_to_coords` there

### CODE REVIEW

**Review by rusheb — CHANGES_REQUESTED** [review]
Nice changes. Few comments, mostly nits

File: maze_transformer/evaluation/path_evals.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+MazePath: TypeAlias = NDArray["node x_y_pos", int]
+PathEvalFunction = Callable[[LatticeMaze, MazePath, MazePath], float]
+
+
+def path_as_segments_iter(path: MazePath) -> Iterable[tuple]:
+    """
+    Iterate over the segments of a path.
+    """
+    i: int
+    n_s: Coord | CoordTup
```
Comment (luciaquirke) [inline maze_transformer/evaluation/path_evals.py:19]: could we rename n_s and n_e to be more meaningful? is this node south and node east 🤔

File: maze_transformer/evaluation/path_evals.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    n_e: Coord | CoordTup
+    for i, n_s in enumerate(path[:-1]):
+        n_e = path[i + 1]
+        yield (n_s, n_e)
```
Comment (luciaquirke) [inline maze_transformer/evaluation/path_evals.py:23]: remove redundant parenthesis

Comment (luciaquirke) [inline maze_transformer/evaluation/path_evals.py:23]:
```suggestion
        yield n_s, n_e
```

File: maze_transformer/evaluation/path_evals.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+
+class PathEvals:
+    """array path based eval functions. first path is always the "ground truth" path"""
+
+    @staticmethod
+    def node_overlap(
```
Comment (luciaquirke) [inline maze_transformer/evaluation/path_evals.py:30]: These static methods have unused input parameters which can be removed

File: maze_transformer/evaluation/path_evals.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        maze: LatticeMaze, solution: MazePath, prediction: MazePath, /
```
Comment (luciaquirke) [inline maze_transformer/evaluation/path_evals.py:31]:
```suggestion
        solution: MazePath, prediction: MazePath, /
```

File: maze_transformer/evaluation/path_evals.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    ) -> float:
+        """fraction of the connections in prediction which actually connect nodes that are adjacent on the lattice, ignoring if they are adjacent on the maze"""
+
+        return PathEvals.num_connections_adjacent_lattice(
+            maze, solution, prediction
+        ) / len(prediction)
+
+    @staticmethod
+    def num_connections_adjacent(
+        maze: LatticeMaze, solution: MazePath, prediction: MazePath, /
```
Comment (luciaquirke) [inline maze_transformer/evaluation/path_evals.py:75]:
```suggestion
        maze: LatticeMaze, prediction: MazePath, /
```

File: maze_transformer/evaluation/path_evals.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    ) -> float:
+        """fraction of connections in prediction which are are valid paths on the maze"""
+
+        return PathEvals.num_connections_adjacent(maze, solution, prediction) / len(
+            prediction
+        )
+
+    @classmethod
+    def all_functions(cls) -> dict[str, PathEvalFunction]:
```
Comment (luciaquirke) [inline maze_transformer/evaluation/path_evals.py:100]: dumb question but what is this for?

File: maze_transformer/evaluation/path_evals.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        excluded = ["all_functions"]
+
+        return {
+            **{
+                name: func
+                for name, func in cls.__dict__.items()
+                if not name.startswith("_") and name not in excluded
+            }
+        }
```
[the remark this answers is not in the document]
Comment (valedan) [reply maze_transformer/evaluation/path_evals.py:109]: Yep this is a fair concern. Enforcing types more strongly would help for sure. 

If there's a more elegant solution I don't know what it is - I did think this code was a bit concerning when I moved it in here. But there is a valid need here for us to be able to easily get a list of all path eval functions. I didn't want to just add some constant listing them somewhere because then there would be 2 places to update any time we change something. 

If there's a more Pythonic way of doing this I'm definitely open to it!

File: maze_transformer/evaluation/path_evals.py (hunk 6 above)
Comment (valedan) [reply maze_transformer/evaluation/path_evals.py:100]: I'm fairly new to Python, so if you find anything on this I'd like to know too. 🙂

File: maze_transformer/evaluation/path_evals.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (valedan) [reply maze_transformer/evaluation/path_evals.py:109]: I wonder if I could add some checks to this function to make sure everything being returned adheres to the correct interface.

File: tests/unit/maze_transformer/evaluation/test_eval_model.py
[the remark this answers is not in the document]
Comment (valedan) [reply tests/unit/maze_transformer/evaluation/test_eval_model.py:1]: Ehh unit vs integration describes what is being tested, not how long the test takes. I would consider this a unit test because it is just directly testing a single function. An integration test would test a complex workflow with multiple functions working together - like our scripts tests.

In general unit tests should be quick to run, and I agree that it is a smell for a unit test to take 9s. But I think that will be resolved as part of #70.

File: maze_transformer/generation/constants.py
[the remark this answers is not in the document]
Comment (valedan) [reply maze_transformer/generation/constants.py:1]: I needed to move a few of them to avoid circular imports, and moved the rest so they'd all be in the same place.

File: maze_transformer/evaluation/path_evals.py (hunk 4 above)
Comment (valedan) [reply maze_transformer/evaluation/path_evals.py:31]: Yeah all these suggestions to remove unused args won't work unless we significantly change the approach here. The path evals all need to have the same interface because they're called in a for loop with the same inputs.

File: maze_transformer/generation/constants.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    target_start="<TARGET_START>",
+    target_end="<TARGET_END>",
+    origin_start="<ORIGIN_START>",
+    origin_end="<ORIGIN_END>",
+    path_start="<PATH_START>",
+    path_end="<PATH_END>",
+    connector="<-->",
+    adjacency_endline=";",
+    padding="<PADDING>",
+)
```
[the remark this answers is not in the document]
Comment (valedan) [reply maze_transformer/generation/constants.py:20]: I'm not very familiar with Python enums - what would be the advantage over a dict?

(Also yeah probably out of scope)

File: maze_transformer/utils/token_utils.py
[the remark this answers is not in the document]
Comment (valedan) [reply maze_transformer/utils/token_utils.py:1]: Hmm, I like that idea. But it would be a huge change, definitely out of scope here. Everything that currently deals with tokens would need to be updated. 

We've discussed before how we'll need to rethink what abstractions we're using for tokenization once we add additional tokenization schemes - maybe we can think about this then?

File: maze_transformer/utils/token_utils.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+
+def tokens_between(tokens: list[str], start_value: str, end_value: str) -> list[str]:
+    start_idx = tokens.index(start_value) + 1
+    end_idx = tokens.index(end_value)
+
+    return tokens[start_idx:end_idx]
+
+
+def get_adjlist_tokens(tokens: list[str]) -> list[str]:
```
[the remark this answers is not in the document]
Comment (valedan) [reply maze_transformer/utils/token_utils.py:13]: No, it's used in #136

File: maze_transformer/utils/token_utils.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
+import typing
+
+from maze_transformer.generation.constants import SPECIAL_TOKENS
+
+
+def tokens_between(tokens: list[str], start_value: str, end_value: str) -> list[str]:
```
[the remark this answers is not in the document]
Comment (valedan) [reply maze_transformer/utils/token_utils.py:6]: > end_idx < start_idx
Good call, adding this.

> tokens which are repeated in the sequence
I don't think this should be an error, I think the current behaviour of just using the first occurrence is what we want. For a token list that includes a model-generated solution, there will often be random special tokens at the end that are duplicates. 

> tokens which don't exist in the sequence
This is already handled - `index` will raise a `ValueError` if the value is not present.

File: maze_transformer/utils/token_utils.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+def get_adjlist_tokens(tokens: list[str]) -> list[str]:
+    return tokens_between(
+        tokens, SPECIAL_TOKENS["adjlist_start"], SPECIAL_TOKENS["adjlist_end"]
+    )
+
+
+def get_path_tokens(tokens: list[str]) -> list[str]:
+    """The path is considered everything from the first path coord to the end of the list, including the path_end token (ie everything we are asking the model to predict)"""
+    start_idx = tokens.index(SPECIAL_TOKENS["path_start"]) + 1
+    return tokens[start_idx:]
```
[the remark this answers is not in the document]
Comment (valedan) [reply maze_transformer/utils/token_utils.py:22]: Hmm I'm torn here. I do see your point, and it would be elegant for everything in here to use `tokens_between`. And if `start_value` was also optional, it would generalize to `get_tokens_up_to_path_start` too. 

But I'm hesitant for a couple of reasons
- I don't want to add too much complexity to `tokens_between`
- It's called tokens_**between** - that doesn't really fit with optional start or end values
- `get_path_tokens` and `get_tokens_up_to_path_start` are fairly simple right now - the benefit of switching them to use `tokens_between` doesn't seem that big.

Right now I'm leaning towards leaving this as is. What do you think?

File: maze_transformer/utils/token_utils.py [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+            output.append(mazedata_cfg.token_node_map[tk])
+        else:
+            if when_noncoord == "skip":
+                continue
+            elif when_noncoord == "include":
+                output.append(tk)
+            elif when_noncoord == "except":
+                raise ValueError(f"token '{tk}' at {idx = } is not a coordinate")
+            else:
+                raise ValueError(f"invalid value for {when_noncoord = }")
```
[the remark this answers is not in the document]
Comment (valedan) [reply maze_transformer/utils/token_utils.py:65]: Lol good point, nice to have a chance to use it 👍

File: tests/unit/maze_transformer/utils/test_token_utils.py [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+        "<ADJLIST_END>",
+        "<ORIGIN_START>",
+        "(1,0)",
+        "<ORIGIN_END>",
+        "<TARGET_START>",
+        "(1,1)",
+        "<TARGET_END>",
+        "<PATH_START>",
+        "(1,0)",
+    ]
```
[the remark this answers is not in the document]
Comment (valedan) [reply tests/unit/maze_transformer/utils/test_token_utils.py:97]: Yep seems reasonable. 👍

**Review by rusheb — COMMENTED** [review]
Removing change request by commenting

File: maze_transformer/evaluation/path_evals.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mivanit) [reply maze_transformer/evaluation/path_evals.py:107]: Adding the following checks might help:
- `isinstance(func, typing.Callable)`
- `isinstance(func, types.FunctionType)` to [check it's a function type](https://stackoverflow.com/questions/8727059/python-check-if-method-is-static)
- `typing.get_type_hints(func)["return"] is float`

`typing.get_type_hints(func)` can also help us check if the argument names & types are correct

a few other notes:
- we probably want to wrap the calls to these functions in a `try ... catch` since we dont want the whole training script to crash if an eval fails. definitely print warnings though, and have a way to make it crash during integration tests. maybe this is already being done?
- is there a reason we are unwrapping the dict only to wrap it again? (maybe this is my fault haha)
- do we want to perform some type checks when loading the module, rather than waiting for `all_functions` to be called elsewhere? this should definitely throw an exception if it finds any functions which dont fit the criteria.
- should we rename `all_functions` to `_all_functions`, thereby excluding it automatically (or maybe move out of the class)? hopefully this discourages adding non-evaluation functions to this class

File: maze_transformer/evaluation/path_evals.py (hunk 6 above)
Comment (mivanit) [reply maze_transformer/evaluation/path_evals.py:98]: the python docs are relatively unhelpful with this afaik, I found these two helpful SO threads:
- https://stackoverflow.com/questions/48029249/python-dict
- https://stackoverflow.com/questions/19907442/explain-dict-attribute

the tl;dr is that it contains all attributes of an object, but all the built-in ones like `__module__`, `__doc__`, and `__dict__` itself start with `__` to signify that they are "magic" methods. Since we only inherit from `object`, there shouldn't be anything other than a built-in in the class.

Starting with `_` is [convention in python](https://peps.python.org/pep-0008/#descriptive-naming-styles) for "this is a private method" or otherwise "don't use this unless you know what you're doing".

> Tangent: the PEP style guide says never to invent double-underscore attributes and only use built-ins, but I may be guilty of doing this sometimes 😢

File: maze_transformer/generation/constants.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mivanit) [reply maze_transformer/generation/constants.py:20]: Agree that this would be a good idea but also agree that it might be out of scope. Enums would mean that:
- type checkers/linters would let us know that we are doing so
- you'd get suggestions in your IDE when trying to get `SPECIAL_TOKENS.adjlist_start`, instead of having to index into the dict and remember the correct name

File: maze_transformer/evaluation/path_evals.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (luciaquirke) [reply maze_transformer/evaluation/path_evals.py:107]: I agree with Michael that the dictionary seems to be being unnecessarily unpacked and repacked. That this syntax slipped through is a sign that no one understands the syntax, which means **we shouldn't use that syntax**.

We should aim for code that is readable and maintainable rather than code which is clever. In this case I think you need to update the eval functions to remove unused parameters. Then you need to update all_functions to call the eval functions one by one (by name) and return their results in a dictionary. The method callsite will need to be updated.

Underscores in Python variables are used to indicate private methods so we shouldn't use it to name a public method.

**Review by luciaquirke — APPROVED** [review]
Ready to merge once the all_functions method has been updated. Update is looking awesome, keen to get this in 🙏

File: maze_transformer/evaluation/path_evals.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mivanit) [reply maze_transformer/evaluation/path_evals.py:107]: Agree that we shouldn't use underscores since it's a public method, and agree that we should aim for readable and maintainable over cleverness.

The best solution here is probably to have `EXCLUDED_FUNCTIONS` as a class variable of the path evals class, rather than a private thing defined anew every time we run `all_functions`. We should also move out the "is this a valid path_eval function" check into a separate function. I'll take a crack at this.

File: maze_transformer/evaluation/path_evals.py (hunk 6 above)
Comment (mivanit) [reply maze_transformer/evaluation/path_evals.py]: sorry if I'm stepping on your toes Dan, I took care of this in #[bf407c5](https://github.com/AISC-understanding-search/maze-transformer/pull/134/commits/bf407c599a330c87c577e133d84ddb3fb748a9bf)

**Comment by valedan** [comment]
Okay I've spent probably an unreasonable amount of time looking into this `all_functions` thing over the past day. I think I have something that I'm moderately happy with. I've removed `all_functions` completely and gone with a decorator that can be used to make arbitrary method dicts. 

A few benefits here:
- Evals are just a class property now, no need to call a method to get them.
- This decorator is easily reusable for other cases where we need a method dict, for example logit-based evals.
- I've removed unused args from all the eval functions, and added a `**_` to allow unused kwargs so that the `evaluate_model` approach still works.

There is a little bit of complexity with some of the type-system wrangling I had to do here, but I think it may be unavoidable. 

@mivanit Thanks for taking a stab at this! I didn't end up using your approach because it was getting a bit too complex, and I don't think we need the `assert`s because that should be taken care of by the function signatures (required params) and type system.

Let me know if this is okay! Really hope I can merge today 😄

**Review by mivanit — APPROVED** [review]
This is a really clean solution! I like this a lot better than what we had before. The type checking I added was mostly just a stopgap until we actually have proper type checking working.

I'd have probably named `evals` as `PATH_EVALS` or something, but that's extremely minor.

Really good work on this!

**Comment by rusheb** [comment]
> gone with a decorator that can be used to make arbitrary method dicts.

Very cool!!
